File: src/main/java/com/rephilo/luandun/leetcode/_058/Solution.java
package com.rephilo.luandun.leetcode._058;

/**
 * 058
 *
 * @author rephilo
 */
public class Solution {
    /**
     * 不知道这个题要干什么，但是先用split试一下
     *
     * @param s
     * @return
     */
    public static int lengthOfLastWord1(String s) {
        if (s == null || s.length() == 0 || " ".equals(s)) {
            return 0;
        }
        String[] words = s.split(" ");
        if (words.length > 0) {
            return words[words.length - 1].length();
        }

        return 0;
    }

    /**
     * indexOf没写出来，用循环试试
     *
     * @param s
     * @return
     */
    public static int lengthOfLastWord2(String s) {
        int result = 0;

        for (int i = s.length() - 1; i >= 0; i--) {
            if (s.charAt(i) != ' ') {
                result++;
            }

            if (s.charAt(i) == ' ' && result > 0) {
                break;
            }
        }

        return result;
    }
}

File: src/main/scala/com/rephilo/luandun/scala/leetcode/_058/Solution.scala
package com.rephilo.luandun.scala.leetcode._058

import scala.annotation.tailrec

/**
  * 058
  *
  * @author rephilo
  */
object Solution {

  /**
    * split
    *
    * @param s
    * @return
    */
  def lengthOfLastWord1(s: String): Int = {
    val array = s.split(" ")
    if (array.nonEmpty) {
      return array(array.length - 1).length
    }

    0
  }

  /**
    * 循环 case 这里写的不好，挺乱的
    *
    * @param s
    * @return
    */
  def lengthOfLastWord2(s: String): Int = {
    @tailrec
    def calc(s: String, index: Int, result: Int): Int = {
      if (index < 0) {
        return result
      }
      val currChar = s.charAt(index)
      currChar match {
        case ' ' =>
          if (result > 0) {
            return result
          }
          calc(s, index - 1, result)
        case _ => calc(s, index - 1, result + 1)
      }
    }

    calc(s, s.length - 1, 0)
  }
}
